[quant] Add FixedQParamsFakeQuantize module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45538 [quant] Add FixedQParamsFakeQuantize module**

Summary:
This is used to simulate fake quantize operation for ops with fixed quantization parameters
e.g. hardsigmoid

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24004795](https://our.internmc.facebook.com/intern/diff/D24004795)

jerryzh168 committed:
[quant] Add FixedQParamsFakeQuantize module

Summary:
This is used to simulate fake quantize operation for ops with fixed quantization parameters
e.g. hardsigmoid

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add FixedQParamsFakeQuantize module"


Summary:
This is used to simulate fake quantize operation for ops with fixed quantization parameters
e.g. hardsigmoid

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24004795](https://our.internmc.facebook.com/intern/diff/D24004795)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add FixedQParamsFakeQuantize module"


Summary:
This is used to simulate fake quantize operation for ops with fixed quantization parameters
e.g. hardsigmoid

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24004795](https://our.internmc.facebook.com/intern/diff/D24004795)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add FixedQParamsFakeQuantize module"


Summary:
This is used to simulate fake quantize operation for ops with fixed quantization parameters
e.g. hardsigmoid

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24004795](https://our.internmc.facebook.com/intern/diff/D24004795)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fake_quantize.py:
```diff
@@ -2,7 +2,43 @@
+
+    with_args = classmethod(_with_args)
+
+# TODO: inherit from FakeQuantizeBase
```
thoughts on an ETA for this? Would be great if the temporary patch did not live in master for long.

jerryzh168 replied:
will do this in 1-2 month

vkuzo commented on torch/quantization/fake_quantize.py:
```diff
@@ -61,6 +97,10 @@ def __init__(self, observer=MovingAverageMinMaxObserver, quant_min=0, quant_max=
+            'Only per channel and per tensor quantization are supported in fake quantize' + \
+            ' got qscheme: ' + str(self.qscheme)
 
     @torch.jit.export
```
should the options to enable/disable fake_quant/observer go on the base class too?

jerryzh168 replied:
I think ideally these should be removed

jerryzh168 replied:
but as a middle step, it should live in base class as well.

vkuzo commented on torch/quantization/fake_quantize.py:
```diff
@@ -137,10 +177,51 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
                                             dtype=torch.quint8, qscheme=torch.per_tensor_affine, reduce_range=True)
 default_weight_fake_quant = FakeQuantize.with_args(observer=MovingAverageMinMaxObserver, quant_min=-128, quant_max=127,
                                                    dtype=torch.qint8, qscheme=torch.per_tensor_symmetric, reduce_range=False)
+default_symmetric_fixed_qparams_fake_quant = FixedQParamsFakeQuantize.with_args(
+    scale=2.0 / 256.0, zero_point=128, dtype=torch.quint8)
+default_affine_fixed_qparams_fake_quant = FixedQParamsFakeQuantize.with_args(
+    scale=1.0 / 256.0, zero_point=0, dtype=torch.quint8, quant_min=-128, quant_max=127)
```
are these tied to a specific function? If so, maybe we could include the function name in the var name, since the range depends on the input function?

jerryzh168 replied:
there are multiple functions that might use these, the mapping will live in a different place.

jerryzh168 commented:
this is not needed anymore, we plan to observe the output for all the ops that inherits quantization parameters (e.g. average pool) and ops that currently have fixed quantization parameters (e.g. sigmoid).

jerryzh168 committed:
Update on "[quant] Add FixedQParamsFakeQuantize module"


Summary:
This is used to simulate fake quantize operation for ops with fixed quantization parameters
e.g. hardsigmoid

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24004795](https://our.internmc.facebook.com/intern/diff/D24004795)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add FixedQParamsFakeQuantize module"


Summary:
This is used to simulate fake quantize operation for ops with fixed quantization parameters
e.g. hardsigmoid

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24004795](https://our.internmc.facebook.com/intern/diff/D24004795)

[ghstack-poisoned]
